gusb/gusb-self-test.c

```diff
@@ -46,8 +46,15 @@ gusb_device_func (void)
 	g_assert_cmpint (array->len, >, 0);
 	device = G_USB_DEVICE (g_ptr_array_index (array, 0));
 
-	g_assert_cmpint (g_usb_device_get_vid (device), >, 0x0000);
-	g_assert_cmpint (g_usb_device_get_pid (device), >, 0x0000);
+	/* Root hubs on FreeBSD have vid and pid set to zero */
```

Comment:
That seems quite unusual. Is there a technical reason the VID/PID is set to zero? Mine are set as "1d6b" i.e. from the Linux Foundation -- I think other session software is going to get confused if the VID and PID are zero too.

Reply:
Hi,

I'm not aware of any USB.org specifications preventing a VID or PID of zero. If using a VID or PID of zero is clever or allowed per USB.org membership, is another matter. Crashing a daemon because the VID or PID is zero sounds like asking for trouble. This chunk seems right.

--HPS

Reply:
Right, nothing is going to crash because the VID/PID is zero, it's just not expected. Whilst I concede a PID of zero is probably valid (and some devices in usb.ids seem to have this...) I still don't think using the vendor ID of zero is correct.

Reply:
Hi,

Having a valid VID is more important when the device uses vendor specific classes. Please keep in mind that not all USB devices which libusb see are real devices!!

Virtual USB devices for example, inside QEMU or VirtualBOX or through remote desktop, might simply use a VID of zero, due to lack of a proper USB.org membership, implementing fully functional devices based in interface class information. Or there might be USB filters in place, avoiding information leakage via USB.

Also keep in mind that under FreeBSD, libusb is "symmetric" and sees both target mode and host mode USB devices. Does libgusb support a mode attribute on each USB device, telling if it is a target or host mode device?